BaikalDataGenerator/Source/object_loader.cpp

```diff
@@ -0,0 +1,390 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#include "object_loader.h"
+
+#include "Rpr/RadeonProRender.h"
+#include "Rpr/WrapObject/Materials/MaterialObject.h"
+
+#include "BaikalIO/image_io.h"
+#include "BaikalIO/material_io.h"
+#include "BaikalIO/scene_io.h"
+
+#include "utils.h"
+
+#include "XML/tinyxml2.h"
+
+#include <iostream>
+#include <unordered_map>
+#include <utility>
+
+
+// validation checks helper macroses to reduce copy paste
+#define ASSERT_PATH(file_name) \
+    if (file_name.empty()) \
+    { \
+        THROW_EX("Missing: " << file_name.string()) \
+    }
+
+#define ASSERT_XML(file_name) \
+    if (file_name.extension() != ".xml") \
+    { \
+        THROW_EX("Not and XML file: " << file_name.string()) \
+    }
+
+#define ASSERT_FILE_EXISTS(file_name) \
+    if (!std::filesystem::exists(file_name)) \
+    { \
+        THROW_EX("File not found: " << file_name.string()) \
+    } \
+
+namespace {
+
+const std::unordered_map<std::string, LightObject::Type> kLightTypesMap
+{
+    { "point", LightObject::Type::kPointLight },
+    { "spot", LightObject::Type::kSpotLight },
+    { "direct", LightObject::Type::kDirectionalLight },
+    { "ibl", LightObject::Type::kEnvironmentLight },
+};
+
+void ValidateConfig(const AppConfig& config)
+{
+    // validate input config
+    ASSERT_PATH(config.camera_file);
+    ASSERT_PATH(config.light_file);
+    ASSERT_PATH(config.spp_file);
+    ASSERT_PATH(config.scene_file);
+    ASSERT_PATH(config.output_dir);
+
+    // validate extensions
+    ASSERT_XML(config.camera_file)
+    ASSERT_XML(config.light_file)
+    ASSERT_XML(config.spp_file)
+
+    // validate that files really exists
+    ASSERT_FILE_EXISTS(config.camera_file)
+    ASSERT_FILE_EXISTS(config.light_file)
+    ASSERT_FILE_EXISTS(config.spp_file)
+    ASSERT_FILE_EXISTS(config.scene_file)
+
+    if (!std::filesystem::is_directory(config.output_dir))
+    {
+        THROW_EX("Not a directory: " << config.output_dir.string())
+    }
+
+    if (config.split_num == 0)
+    {
+        THROW_EX("'split_num' should be positive");
+    }
+
+    if (config.split_idx >= config.split_num)
+    {
+        THROW_EX("'split_idx' must be less than split_num");
+    }
+}
+
+std::pair<std::size_t, std::size_t> GetSplitByIdx(std::size_t total_num,
```

Comment:
I would return a struct with human readable field names, pair is a bit confusing on the client side

Reply:
or at least using range = std::pair<size_t, size_t>, and range.begin, range.end would make much more sense